File: ruby/find_first_duplicate.rb
#my test cases
#input => output 
# [1,2,2,3] => 2
# [1,1] => 1
# [1, 3, 4, 4] => 4


def find_first_duplicate(arr)
  # Look/iterate through an array to find the first case of a duplicate number. If there is no duplicate the output 
  # should be -1. 
  # Tool box: includes..? uniquness? .filter? .uniq
  new_array = []
  arr.uniq!.
 
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  const numbersSeen = []
  // type your code here
  //could have an array that stores, or an object. 
  for(let i = 0; i < arr.length; i ++){

    if (numbersSeen.includes(element)){
      return element
      } else {
        numbersSeen.push(element)
      }
    }

    return -1

  }
  // const numbersSeen = []
  // result = -1
  // arr.forEach(element => { 

  //  if (numbersSeen.includes(element)) {
  //   result = element 
  //     return;
  //   } else {

  //     numbersSeen.push(element)
  //   }
  //   // has this element been encountered before.
  //   //includes

    
  // });
  // return result;

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;

// Please add your pseudocode to this file
// And a written explanation of your solution
